### PR TITLE
[ROCm] [Fused Moe EP] Use binary expert mask for aiter fused moe kernel

### PR DESCRIPTION
**Pull request by ZhiweiYan-96**
## Purpose

Quark Fused Moe meets acc issue in DeepSeek-R1 Expert Parallelism. This PR fixes it through setting the right expert mask with format required by aiter kernel. The aiter kernel requires the mask to be a bitmask. 1 means the expert is on current rank, while 0 is not.

## Test Plan

Server Launch

```bash
export VLLM_USE_V1=1
# export VLLM_LOGGING_LEVEL=DEBUG
export VLLM_RPC_TIMEOUT=1800000
export VLLM_USE_TRITON_FLASH_ATTN=1
export VLLM_ROCM_USE_AITER=1
export VLLM_ROCM_USE_AITER_MHA=0
export VLLM_ROCM_USE_AITER_MLA=0
export VLLM_ROCM_USE_AITER_MOE=1
export VLLM_ROCM_USE_AITER_TRITON_ROPE=1
export VLLM_DISABLE_COMPILE_CACHE=1
export VLLM_ROCM_USE_AITER_FP4_ASM_GEMM=0
export VLLM_ROCM_USE_AITER_FUSION_SHARED_EXPERTS=0 

export TRITON_HIP_ASYNC_COPY_BYPASS_PERMUTE=1
export TRITON_HIP_USE_ASYNC_COPY=1
export TRITON_HIP_USE_BLOCK_PINGPONG=1
export TRITON_HIP_ASYNC_FAST_SWIZZLE=1
export NCCL_DEBUG=WARN
export AMDGCN_USE_BUFFER_OPS=1
export SAFETENSORS_FAST_GPU=1

model_path=amd/DeepSeek-R1-MXFP4-Preview
echo "running $model_path"

vllm serve $model_path \
  --host localhost \
  --port 9000 \
  --data-parallel-size 8 \
  --enable-expert-parallel \
  --max-num-batched-tokens 32768 \
  --trust-remote-code \
  --no-enable-prefix-caching \
  --disable-log-requests \
  --compilation-config '{"cudagraph_mode": "FULL_AND_PIECEWISE"}' \
  --gpu_memory_utilization 0.8 \
  --async-scheduling \
  --load-format fastsafetensors \
  --block-size 16 \
  --seed 123 2>&1 | tee log.server.log &
```

Bechmark

```
addr=localhost
port=9000
url=http://${addr}:${port}/v1/completions
model=amd/DeepSeek-R1-MXFP4-Preview
bs=50
task=gsm8k

echo "url=${url}"
echo "model=${model}"
echo "task=${task}"

lm_eval \
    --model local-completions \
    --tasks ${task} \
    --model_args model=${model},base_url=${url} \
    --batch_size ${bs} \
    --seed 123 \
    2>&1 | tee log.lmeval.log
```

## Test Result

Following are the benchmark results on dataset gsm8k

* FULL_AND_PIECEWISE mode
image

* Eager mode

image

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

**Comment by ZhiweiYan-96** [comment]
@zejunchen-zejun  @gshtras @wuhuikx @tjtanaa  Could you please help review? Thanks.

File: vllm/_aiter_ops.py [hunk 1]
```diff
@@ -94,6 +94,9 @@ def _rocm_aiter_fused_moe_impl(
     activation = ActivationType(activation_method)
     quant_type = QuantType(quant_method)
 
+    assert torch.all((expert_mask == 0) | (expert_mask == 1)), (
```
Comment (zejunchen-zejun) [inline vllm/_aiter_ops.py]: If EP=1, the expert mask could be None?

Comment (ZhiweiYan-96) [reply vllm/_aiter_ops.py]: Yes, gemini also find it, it's a bug here.

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 2]
```diff
@@ -520,6 +520,28 @@ def __init__(
...
+                    torch.tensor(
+                        [
+                            self.local_num_experts + i
+                            for i in range(self.num_fused_shared_experts)
+                        ],
+                        dtype=torch.int32,
+                    ),
+                ),
+                dim=0,
+            )
```
[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply vllm/model_executor/layers/fused_moe/layer.py]: it's a customized handling for rocm aiter shared fused moe kernels. The map generated by `determine_expert_map` will not be handled correctly.

File: vllm/_aiter_ops.py [hunk 3]
```diff
@@ -94,6 +94,9 @@ def _rocm_aiter_fused_moe_impl(
...
+        "Aiter Fused MoE kernel only supports expert_map with 0 and 1s."
+    )
```
[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply vllm/_aiter_ops.py]: good catch

File: vllm/model_executor/layers/fused_moe/layer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tjtanaa) [reply vllm/model_executor/layers/fused_moe/layer.py]: The `return_expert_mask:bool` is specifically designed for rocm_aiter_fused_moe , can you fix that instead so that the logic extend to fix this use case?

[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply vllm/model_executor/layers/fused_moe/layer.py]: oh, got your point. May need to figure out why former mask is wrong.

**Comment by heheda12345** [comment]
CC @tjtanaa

File: vllm/model_executor/layers/fused_moe/layer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply vllm/model_executor/layers/fused_moe/layer.py]: Turns out that, the logic is redundant. Only passe self.expert_map to aiter fused moe is required. I removed the redundant code.

**Review by tjtanaa — APPROVED** [review]
LGTM.
